WordPress/gutenberg
# Block Bindings: Use post meta label from `register_meta` in block bindings workflows

## What?
Fixes https://github.com/WordPress/gutenberg/issues/65066

This pull request is built on top of https://github.com/WordPress/gutenberg/pull/64072 because it reuses the `getRegisteredPostMeta` resolver.

In this PR, I'm experimenting with using the label exposed there to consume it in the places where we are using block bindings. These are the use cases covered so far:

**In the attributes panel**

| Before | After |
|----------|----------|
| ![Screenshot 2024-09-05 at 19 30 56](https://github.com/user-attachments/assets/21f6d545-e6c5-42f6-80c0-026dfe52e277) | ![Screenshot 2024-09-05 at 19 29 44](https://github.com/user-attachments/assets/24266774-5066-4f41-8d7a-04bc47b004df) |

You can appreciate the difference both in the attributes section and in the dropdown.

**In rich text placeholders for empty attributes**
**When the user can edit**

| Before | After |
|----------|----------|
| ![Screenshot 2024-09-05 at 19 25 53](https://github.com/user-attachments/assets/9c093275-b003-4f7b-bd48-e4d5da87573d) | ![Screenshot 2024-09-05 at 19 27 24](https://github.com/user-attachments/assets/6888be19-5a87-4a77-881b-e518ed8b93da) |



**When the user can not edit**

| Before | After |
|----------|----------|
| ![Screenshot 2024-09-05 at 19 26 21](https://github.com/user-attachments/assets/4c2636a7-334e-472d-ade1-fb75b2814c0f) | ![Screenshot 2024-09-05 at 19 27 46](https://github.com/user-attachments/assets/15a1adbf-a8ac-4b93-a416-d80ae491b69c) |

**When no label is provided, it still works as before**

## Why?
Right now, we are relying on the key, which is a bit technical. A recurrent feedback received is that it'd be nice to have a human-readable label.

## How?
The main change done here is changing slightly how the `getFieldsList` API works (still private). Until now, it was returning an object of key: value. Now, I explore returning key: args. And including in those args both the label and the value. That way, it provides flexibility for each source. In the case of post meta, we get the value using `getRegisteredPostMeta` resolver.

Return from `getFieldsList` API:

```js
// Before
{
    custom_field_1: "Value 1",
    custom_field_2: "Value 2"
}
// After

```js
{
    custom_field_1: {
        label: "Label 1",
        value: "Value 1",
    },
    custom_field_2: {
        label: "Label 2",
        value: "Value 2",
    },
}
```

## Testing Instructions
1. Register some meta fields, with and without a label.
```
	register_meta(
		'post',
		'abel_field',
		array(
			'label'   => 'My label',
			'show_in_rest'   => true,
			'single'         => true,
			'type'           => 'string',
			'default'        => '',
		)
	);

	register_meta(
		'post',
		'no_label_field',
		array(
			'show_in_rest'   => true,
			'single'         => true,
			'type'           => 'string',
			'default'        => '',
		)
	);
```
2. Add a paragraph and check the different workflows.
